### PR TITLE
HUE-9549 [SQLAlchemy] API - Support explain

### PR DESCRIPTION
**Pull request by agl29**

### CODE REVIEW

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py [hunk 1]
```diff
@@ -219,6 +219,36 @@ def test_check_status(self):
...
+          'key_len': None,
+          'ref': None,
+          'rows': 1,
+          'filtered': None,
+          'Extra': 'No tables used'
+        }
+        
+        engine.execute = Mock(return_value=result)
+
+        response = SqlAlchemyApi(self.user, self. interpreter).explain(notebook, snippet)
```
Comment (agl29) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py:247]: Don't know why in this response['explanation'] is empty.

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy.py [hunk 2]
```diff
@@ -298,6 +298,13 @@ def explain(self, notebook, snippet):
     if statement:
       if self.options['url'].startswith('bigquery://'):
         explanation = ''
+      else:
+        result = connection.execute('EXPLAIN '+ statement)
+  
+        explanation = [dict(r) for r in result]
+        explanation = "\n".join("{!r}: {!r},".format(k, v) for dictt in explanation for k, v in dictt.items())
```
Comment (romainr) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy.py:305]: Why !r nd not simpler?


https://docs.python.org/3/tutorial/inputoutput.html#the-string-format-method

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py [hunk 3]
```diff
@@ -243,6 +243,29 @@ def test_create_engine_with_impersonation(self):
...
+          
+          execute = Mock(return_value=result)
+          _create_connection.return_value = Mock(
+            execute=execute
+          )
+          notebook = {}
+          snippet = {'statement': 'SELECT 1;'}
+          
+          explanation = [dict(r) for r in result]
+          explanation = "\n".join("{!r}: {!r},".format(k, v) for dictt in explanation for k, v in dictt.items())
```
Comment (romainr) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py:262]: Why this?

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py [hunk 4]
```diff
@@ -243,6 +243,29 @@ def test_create_engine_with_impersonation(self):
       create_engine.assert_called_with('presto://test@hue:8080/hue', pool_pre_ping=True)
 
 
+  def test_explain(self):
+
+    with patch('notebook.connectors.sql_alchemy.SqlAlchemyApi._create_connection') as _create_connection:
+      with patch('notebook.connectors.sql_alchemy.SqlAlchemyApi._create_engine') as _create_engine:
+        with patch('notebook.connectors.sql_alchemy.SqlAlchemyApi._get_session') as _get_session:
+
+          result = [(("id", 1), ("select_type", "SIMPLE"), ("Extra", "No tables used"))]
```
Comment (romainr) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py:252]: Please use the final output of explain, should be just a dict?

```
return {
      'status': 0,	      'status': 0,
      'explanation': explanation,	      'explanation': explanation,
      'statement': statement	      'statement': statement
    }	    }

```

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py [hunk 5]
```diff
@@ -243,6 +243,29 @@ def test_create_engine_with_impersonation(self):
       create_engine.assert_called_with('presto://test@hue:8080/hue', pool_pre_ping=True)
 
 
+  def test_explain(self):
```
Comment (romainr) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py:246]: Super nice!

**Review by romainr — COMMENTED** [review]
Good stuff, let's polish it a bit but the core is great!

File: desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py [hunk 6]
```diff
@@ -243,6 +243,28 @@ def test_create_engine_with_impersonation(self):
...
+          result = [{"id": 1, "select_type": "SIMPLE"}, {"Extra": "No tables used"}]
+          
+          execute = Mock(return_value=result)
+          _create_connection.return_value = Mock(
+            execute=execute
+          )
+          notebook = {}
+          snippet = {'statement': 'SELECT 1;'}
+
+          explanation = "\n".join("{}: {},".format(k, v) for row in result for k, v in row.items())
```
Comment (romainr) [inline desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py:261]: Shouldn't it be a raw string instead of redoing the same code logic?

Comment (agl29) [reply desktop/libs/notebook/src/notebook/connectors/sql_alchemy_tests.py]: done

**Review by romainr — APPROVED** [review]

**Pull request merged**